notes/plutus-core/Bytestrings.md

```diff
@@ -0,0 +1,116 @@
+## Extending bytestring operations in PlutusTx and Plutus Core
+(19th November 2020)
+
+In a [Slack thread](https://input-output-rnd.slack.com/archives/C21UF2WVC/p1599120765047200)
+some time ago, Alex asked about extending support for bytestrings in Plutus Core.
+This document describes the issue and suggests some possible extensions.
+
+
+### Current bytestring support in PLC
+
+Plutus Core currently provides the following operations for its built-in
+bytestring type:
+
+   * `concatenate`
+   * `equalsByteString`
+   * `lessThanByteString`
+   * `greaterThanByteString`
+   * `takeByteString`
+   * `dropByteString`
+   * `sha2_256`
+   * `sha3_256`
+   * `verifySignature`
+
+The `concatenate` function might be better called `appendByteString`.  The PLC
+specification also mentions an `emptyByteString` builtin, but we don't support
+that.  However we do provide concrete syntax for bytestrings:
+`(con bytestring #91ef22)` for example; `(con bytestring #)` represents the empty bytestring.
+
+The problem is that there is no easy way to _construct_ a bytestring in Haskell
+and convert it into a PLC bytestring, or to programatically construct a new
+bytestring inside a Plutus Core program.  The `Data` type supports bytestrings
+and can be used to pass bytestrings into validators, but this isn't very
+general-purpose.
+
+
+### Characters and strings in Plutus Core
+
+Plutus Core also provides built-in character and string types.  These both have
+concrete syntactical representations: `(con char 'q')` and `(con string
+"something")`.  Support for these is very limited: we only have `charToString`
+and `append` (append one string to another); also the PlutusTx compiler can
+compile literal Haskell strings and characters into PLC values of the
+appropriate types (but see below for some complications involving Strings).
+
+However there is a subtlety: Plutus Core actually has _two_ representations of
+Haskell Strings.  The PlutusTx plugin compiles normal Haskell strings into
+Scott-encoded lists of PLC's `char` type ("Scott strings" for short), but there
```

Comment:
Hmmm. I wonder if we could just stop doing that and refuse to support non-overloaded string literals :thinking:

Reply:
I do find it a bit confusing, but on the other hand you get lots of string operations for free.